plugins/thunderbolt/fu-thunderbolt-device.c

```diff
@@ -0,0 +1,645 @@
+	istr = G_INPUT_STREAM (g_file_read (nvmem, NULL, error));
+	if (istr == NULL)
+		return FALSE;
+	controller_fw = g_input_stream_read_bytes (istr,
+						   nr_chunks * FU_TBT_CHUNK_SZ,
+						   NULL, error);
+	if (controller_fw == NULL)
+		return FALSE;
+	if (!fu_firmware_parse (FU_FIRMWARE (firmware),
+				controller_fw,
+				FWUPD_INSTALL_FLAG_NONE,
+				error))
+		return FALSE;
+	*is_native = fu_thunderbolt_firmware_is_native (firmware);
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_can_update (FuThunderboltDevice *self)
+{
+	g_autoptr(GError) nvmem_error = NULL;
+	g_autoptr(GFile) non_active_nvmem = NULL;
+
+	non_active_nvmem = fu_thunderbolt_device_find_nvmem (self, FALSE,
+							     &nvmem_error);
+	if (non_active_nvmem == NULL) {
+		g_debug ("%s", nvmem_error->message);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_get_version (FuThunderboltDevice *self)
+{
+	const gchar *version_raw = NULL;
+	g_auto(GStrv) split = NULL;
+	g_autofree gchar *version = NULL;
+
+	for (guint i = 0; i < 50; i++) {
+		version_raw = fu_udev_device_get_sysfs_attr (FU_UDEV_DEVICE (self),
+							     "nvm_version",
+							     NULL);
+		if (version_raw != NULL)
+			break;
+		g_debug ("Attempt %u: Failed to read NVM version", i);
+		if (errno != EAGAIN)
+			break;
+		g_usleep (TBT_NVM_RETRY_TIMEOUT * 1000);
+	}
+
+	if (version_raw == NULL)
+		return FALSE;
+	split = g_strsplit (version_raw, ".", -1);
+	if (g_strv_length (split) != 2)
+		return FALSE;
+
+	version = g_strdup_printf ("%02x.%02x",
+				   (guint) g_ascii_strtoull (split[0], NULL, 16),
+				   (guint) g_ascii_strtoull (split[1], NULL, 16));
+	fu_device_set_version (FU_DEVICE (self), version);
+	return TRUE;
+}
+
+static void
+fu_thunderbolt_device_check_safe_mode (FuThunderboltDevice *self)
+{
+	g_autoptr(GError) error_local = NULL;
+	g_autofree gchar *test_safe = NULL;
+	g_autofree gchar *safe_path = NULL;
+	/* glib can't return a properly mapped -ENODATA but the
+		* kernel only returns -ENODATA or -EAGAIN */
+	safe_path = g_build_path ("/", self->devpath, "nvm_version", NULL);
+	if (!g_file_get_contents (safe_path, &test_safe, NULL, &error_local) &&
+		!g_error_matches (error_local, G_IO_ERROR, G_IO_ERROR_WOULD_BLOCK)) {
+		g_warning ("%s is in safe mode --  VID/DID will "
+				"need to be set by another plugin",
+				self->devpath);
+		self->safe_mode = TRUE;
+		fu_device_set_version (FU_DEVICE (self), "00.00");
+		fu_device_add_instance_id (FU_DEVICE (self), "TBT-safemode");
+		fu_device_set_metadata_boolean (FU_DEVICE (self), FU_DEVICE_METADATA_TBT_IS_SAFE_MODE, TRUE);
+	}
+}
+
+static void
+fu_thunderbolt_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE (device);
+	fu_common_string_append_kb (str, idt, "Host Controller", self->host);
+	fu_common_string_append_kb (str, idt, "Safe Mode", self->safe_mode);
+}
```

Comment:
Since the FW is being parsed at daemon startup now we may as well also add the following:
* "Is Native"
* Generation

Reply:
Added these.